feat: add ImageGenerationModel Class

mederka opened the pull request:
This PR introduces `ImageGenerationModel` class that allows for the generation of images from text prompts using the diffusion models with support for LoRAs. The inference happens on [fal](https://fal.ai). The key changes are as follows:

- **New Data Models**: Added `ImageGenerationModelImageResult`, `ImageGenerationModelResponse`, `ImageSize`, `ImageSizePreset`, `ImageSizeInput`, and `LoraWeight` data models to handle the image generation process.

- **ImageGenerationModel Class**: This is the main class that uses the above data models to generate images. It includes the following features:
  *Initialisation*: The class can be initialized with a model URL and FAL credentials. If not provided, it defaults to environment variables.
  *Image Generation*: The `generate` and `async_generate` methods allow for the synchronous and asynchronous generation of images from text prompts. These methods support various parameters such as `prompt`, `model_name`, `negative_prompt`, `loras`, `seed`, `image_size`, `num_inference_steps`, `guidance_scale`, `clip_skip`, `model_architecture`, `scheduler`, `image_format`, and `num_images`.
*Error Handling*: The class includes error handling to ensure that fal credentials are provided during initialization.

This new feature enhances the project's capabilities by allowing users to generate images from text prompts, which can be useful in a variety of applications, such as content generation, AI art, etc.

Please review and let me know if there are any concerns or suggestions for improvement.

mederka commented:
# ImageGeneration Model Documentation

## Image generation models

Image generation models can be used to create images from text prompts using a specific diffusion model. The model `ImageGenerationModel` is hosted on a [fal](https://fal.ai)
and accessed using the fal key credentials. This class provides both synchronous and asynchronous methods to generate images.

## Usage

To use the ImageGenerationModel, you need to provide the [fal key credentials](https://www.fal.ai/docs/authentication/key-based). You can get them by logging into the [fal
dashboard](https://fal.ai/dashboard). Credentials be provided as arguments when creating an instance of ImageGenerationModel, or they can be set as environment variables
("FAL_KEY_ID" and "FAL_KEY_SECRET"). If they are not provided in either of these ways, a ValueError will be raised.

### Synchronous generation
```python
from replit.ai.modelfarm import ImageGenerationModel

# Create an instance of the ImageGenerationModel
model = ImageGenerationModel()

# Generate an image from a text prompt
response = model.generate(prompt="A beautiful sunset over the ocean.")

# Print the URL of the generated image
for image in response.images:
    print(image.url)
```

### Asynchronous generation
```python
import asyncio
from replit.ai.modelfarm import ImageGenerationModel

# Create an instance of the ImageGenerationModel
model = ImageGenerationModel()

# Define an asynchronous function to generate an image
async def generate_image():
    response = await model.async_generate(
        prompt="A city skyline at night.",
        model_name="stabilityai/stable-diffusion-xl-base-1.0",
        image_size="landscape_16_9",
        num_inference_steps=50,
        image_format="jpeg",
    )

    # Print the URL of the generated image
    for image in response.images:
        print(image.url)

# Run the asynchronous function
asyncio.run(generate_image())
```

### Image generation with LoRAs
```python
from replit.ai.modelfarm import ImageGenerationModel, LoraWeight

# Initialize the ImageGenerationModel
txt2img_model = ImageGenerationModel()

# Define the LoraWeights
loras = [
    LoraWeight(path="https://civitai.com/api/download/models/135931", scale=0.5),
    LoraWeight(path="https://filebin.net/3chfqasxpqu21y8n/my-custom-lora-v1.safetensors", scale=0.5)
]

# Generate an image
response = txt2img_model.generate(
    prompt="A beautiful sunset over the ocean",
    loras=loras,
    image_size="square_hd",
    num_images=1
)

# Print the URL of the generated image
print(response.images[0].url)
```

## Classes
### ImageGenerationModel

This class is used to generate images from text prompts. It has two methods for generating images: generate and async_generate. Both methods take the same parameters:

    prompt: The text prompt to generate the image from.
    model_name: The URL or HuggingFace ID of the base model to generate the image. Default is "stabilityai/stable-diffusion-xl-base-1.0".
    negative_prompt: A text prompt for details that should not be present in the image.
    loras: A list of LoraWeight instances to use for the image generation.
    seed: A seed for the random number generator. Providing the same seed and prompt will result in the same image.
    image_size: The size of the image to generate. This can be an instance of ImageSize or one of the following presets: "square_hd", "square", "portrait_4_3", "portrait_16_9",
"landscape_4_3", "landscape_16_9". Default is "square_hd".
    num_inference_steps: Number of inference steps. Default is 30.
    guidance_scale: Guidance scale. Default is 7.5.
    clip_skip: Skips part of the image generation process. Default is 0.
    model_architecture: The architecture of the model to use. Can be either "sd" or "sdxl".
    scheduler: Scheduler to use for the image denoising process.
    image_format: The format of the generated image. Can be either "png" or "jpeg". Default is "png".
    num_images: Number of images to generate in one request. Default is 1.

Both methods return an instance of ImageGenerationModelResponse.
### ImageGenerationModelImageResult

This class represents the result of an image generation request. It contains the following attributes:

    url: The URL where the generated image is hosted.
    content_type: The MIME type of the image.
    file_name: The name of the image file.
    file_size: The size of the image file in bytes.
    width: The width of the image in pixels.
    height: The height of the image in pixels.

### ImageGenerationModelResponse

This class represents the response from the image generation model. It contains the following attributes:

    images: A list of ImageGenerationModelImageResult instances, each representing a generated image.
    seed: The seed used for the image generation.

This class also has a class method from_response_json which takes a dictionary (typically the JSON response from the server) and returns an instance of
ImageGenerationModelResponse.

### ImageSize

This class represents the size of the image to be generated. It contains the following attributes:

    width: The width of the image to be generated. Default is 512 pixels. The value must be greater than 0 and less than or equal to 4096.
    height: The height of the image to be generated. Default is 512 pixels. The value must be greater than 0 and less than or equal to 4096.

### LoraWeight

This class represents the LoRA weights used for image generation. It contains the following attributes:

    path: The URL or the path to the LoRA weights.
    scale: The scale of the LoRA weight. This is used to scale the LoRA weight before merging it with the base model. The value must be between 0.0 and 1.0.

burkaygur commented:
Quick typo in docs: "Run on a fal" vs "Run on fal"

chamini2 commented on src/replit/ai/modelfarm/txt2image_model.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        url = f"{self.model_url}"
+        headers = {"Authorization": f"Basic {self.fal_key_id}:{self.fal_key_secret}"}
+        args_dict = locals()
+        del args_dict["self"]
```
not really sure if this is OK. should you clone the dict before modifying it?

chamini2 replied:
are `url` and `headers` included in the `locals()`?

chamini2 replied:
and what about `__name__`, `__doc__` and others

```
>>> locals()
{'__name__': '__main__', '__doc__': None, '__package__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': None, '__annotations__': {}, '__builtins__': <module 'builtins' (built-in)>, 'a': 1}
```

badayvedat replied:
I agree with Matteo, and I think It would be clearer if we only passed the required arguments instead of dumping all of the `locals()`

isidentical replied:
I agree. I would really think we should just pass all the inputs as is and not resort to this (unless this is the convention they are using).

isidentical commented on src/replit/ai/modelfarm/__init__.py:
```diff
@@ -1,6 +1,7 @@
 from .completion_model import CompletionModel as CompletionModel
 from .chat_model import ChatModel as ChatModel
 from .embedding_model import EmbeddingModel as EmbeddingModel
+from .txt2image_model import Txt2ImgModel as Txt2ImgModel
```
Is this the name we want to go with? It seems a bit hard to parse compared to `Text2Image` or even better `ImageGenerationModel` (following the suite of `CompletionModel`, `ChatModel` and `EmbeddingModel`). OpenAI also calls this image generation instead of Text2Image.

isidentical commented on src/replit/ai/modelfarm/structs.py:
```diff
@@ -86,3 +86,68 @@ class EmbeddingModelRequest(BaseModel):
+    images: List[Txt2ImgModelImageResult]
+    seed: int
+
+    @classmethod
```
How does other models handle this nested deserialization problem? E.g. `EmbeddingModelResponse` can contain a list of `Embedding`s but they don't define this method to deserialize.

isidentical commented on src/replit/ai/modelfarm/txt2image_model.py:
```diff
@@ -0,0 +1,149 @@
+    ),
+    "Euler": ("EulerDiscreteScheduler", {}),
+    "Euler A": ("EulerAncestralDiscreteScheduler", {}),
+}
```
This should be just a type now with `Literal["a", "b", "c"]`. The kv details aren't meant to be used publicly.

isidentical commented on src/replit/ai/modelfarm/txt2image_model.py:
```diff
@@ -0,0 +1,149 @@
+from .structs import Txt2ImgModelResponse, LoraWeight, ImageSizeInput
+from typing import Optional, Any, Dict, Literal
+
+DEFAULT_LORA_URL = "https://110602490-lora.gateway.alpha.fal.ai"
```
Shouldn't this be part of https://github.com/fal-ai/replit-ai-python/blob/8af6a3c579367b21e541a9fe7d00040856130de7/src/replit/ai/modelfarm/config.py#L4-L9?

isidentical commented on src/replit/ai/modelfarm/txt2image_model.py:
```diff
@@ -0,0 +1,149 @@
+        prompt: str,
+        model_name: str = "stabilityai/stable-diffusion-xl-base-1.0",
+        negative_prompt: str = "",
+        loras: Optional[list[LoraWeight]] = [],
```
Default here should be none, not a list.

mederka replied:
turns out it's not optional, we have to pass a list, updating the type hint

mederka committed:
feat: add classes to help handle image generation

mederka committed:
feat: add ImageGenerationModel class

This class lets users generate images using fal-hosted model with LoRAs enabled

mederka committed:
test: add image generation tests

chamini2 commented on src/replit/ai/modelfarm/image_generation_model.py:
```diff
@@ -0,0 +1,170 @@
+        Returns:
+            ImageGenerationModelResponse
+        """
+        url = f"{self.model_url}"
+        headers = {"Authorization": f"Basic {self.fal_key_id}:{self.fal_key_secret}"}
+        response = requests.post(
+            url,
+            headers=headers,
+            json={
+                "prompt": prompt,
+                "model_name": model_name,
+                "negative_prompt": negative_prompt,
+                "loras": loras,
+                "seed": seed,
+                "image_size": image_size,
+                "num_inference_steps": num_inference_steps,
+                "guidance_scale": guidance_scale,
+                "clip_skip": clip_skip,
+                "model_architecture": model_architecture,
+                "scheduler": scheduler,
+                "image_format": image_format,
+                "num_images": num_images,
+            },
+        )
+        self._check_response(response)
+        return ImageGenerationModelResponse(**response.json())
```
would it be better if the sync solution just calls the async one and waits for it to finishi? it would be like the 

```
loop = asyncio.get_running_loop()
await loop.run_in_executor(...)
```

we have done internally.


No hard opinions, just an idea.
